whelktool/scripts/cleanups/2022/06/lxl-3880-shuffle-work-titles.groovy

```diff
@@ -0,0 +1,400 @@
+TRANSLATION_OF = "translationOf"
+LANGUAGE = "language"
+ID = '@id'
+
+selectByCollection('bib') { bib ->
+    Map instance = bib.graph[1]
+    def instanceOf = instance.find { it.key == INSTANCE_OF }
+    def shortId = bib.doc.shortId
+
+    if (!instanceOf)
+        return
+
+    if (multipleValues(instanceOf.value)) {
+        multiple.println("$shortId\t$INSTANCE_OF")
+        return
+    }
+
+    Map work = asList(instanceOf.value)[0]
+
+    // Remove superfluous blank languages
+    def worklang = asList(work[LANGUAGE])
+    def whichOne = worklang.findAll { it[ID] in ambiguousLangIds }
+    def blankLang = worklang.find { !it[ID] }
+    if (blankLang) {
+        def mapped = mapBlankLanguages([blankLang], whichOne)
+        if ((worklang - blankLang).toSet() == mapped.toSet()) {
+            worklang.remove(blankLang)
+            blankLangRemoved.println([shortId, blankLang.label, worklang].join('\t'))
+        }
+    }
+
+    def expressionOf = work.find { it.key == EXPRESSION_OF }
+
+    if (expressionOf) {
+        if (multipleValues(expressionOf.value)) {
+            multiple.println("$shortId\t$EXPRESSION_OF")
+            return
+        }
+
+        // Move expressionOf content primarily to translationOf, otherwise to instanceOf
+        def target = work.find { it.key == TRANSLATION_OF } ?: instanceOf
+
+        Map hub = asList(expressionOf.value)[0]
+        if (hub[ID]) {
+            expressionOf.value = loadThing(hub[ID])
+            if (!expressionOf.value) {
+                brokenLinks.println([shortId, hub[ID]].join('\t'))
+            }
+            // Linked uniform work title in expressionOf: move hasTitle to target and then remove the link
+            else if (moveProperty(shortId, HAS_TITLE, expressionOf, target)) {
+                work.remove(EXPRESSION_OF)
+            }
+        }
+        // Move title from local expressionOf to target
+        else if (moveProperty(shortId, HAS_TITLE, expressionOf, target)) {
+            hub.remove(HAS_TITLE)
+            // Move the complement expressionOf.language \ instanceOf.language to instanceOf.language
+            if (hub[LANGUAGE]) {
+                def langsAdded = []
+                asList(hub[LANGUAGE]).each { l ->
+                    asList(l.'@id' ? l : mapBlankLanguages([l], whichOne)).each {
+                        if (!(it in worklang)) {
+                            worklang << it
+                            langsAdded << it
+                        }
+                    }
+                }
+                if (langsAdded) {
+                    work[LANGUAGE] = worklang
+                    extraLang.println([shortId, langsAdded, worklang].join('\t'))
+                }
+                hub.remove(LANGUAGE)
+            }
+            // Move remaining properties (except @type) to the same place as the title
+            hub.removeAll {
+                it.key == '@type' ?: moveProperty(shortId, it.key, expressionOf, target)
```

Comment:
It fails if any of the properties already exists on the target. These cases have been cleaned up manually but there is a possibility that new records keep coming in with this bad structure.

My thinking was that it doesn't matter if it fails, since it will be reported in 'property-already-exists.tsv' (for manual handling) but yeah, maybe it's better to not save the update at all if it partly fails. See [3d2293a](https://github.com/libris/librisxl/pull/1130/commits/3d2293a30d7cbece0b12adbc2bdf9e99c6f7bd7a)